api/lib/application/assessments/assessment-controller.js

```diff
@@ -26,8 +44,80 @@ module.exports = {
     assessmentRepository
       .get(request.params.id)
       .then((assessment) => {
-        const serializedAssessment = assessmentSerializer.serialize(assessment);
-        return reply(serializedAssessment);
+        courseRepository
+          .get(assessment.get('courseId'))
+          .then((course) => {
+
+            const challengePromises = course.challenges.map(challengeId => challengeRepository.get(challengeId));
+
+            Promise.all(challengePromises)
+              .then(challenges => {
+
+                const knowledgeOf = {};
+                const difficultyOf = {};
+                const allKnowledge = {};
+                _.forEach(challenges, challenge => {
+                  if(challenge.knowledge !== undefined) {
+                    challenge.knowledge.forEach(knowledge => allKnowledge[knowledge] = 1);
+                    knowledgeOf[challenge.id] = challenge.knowledge;
+                    difficultyOf[challenge.id] = parseInt(challenge.knowledge[0].slice(-1));
+                  }
+                });
+
+                const acquired = [];
+                const notAcquired = [];
+                const history = [];
+                answerRepository.findByAssessment(assessment.get('id')).then((answers) => {
+
+                  if (answers.length == 0) {
+                    const serializedAssessment = assessmentSerializer.serialize(assessment);
+                    return reply(serializedAssessment);
+                  }
+
+                  const modelAnswers = _.map(answers.models, o => o.attributes);
+
+                  _.forEach(modelAnswers, function(answer) {
+                    if(knowledgeOf.hasOwnProperty(answer.challengeId)) {
+                      const startNode = knowledgeOf[answer.challengeId][0];
+                      if (answer.result == 'ok') {
+                        history.push({diff: difficultyOf[answer.challengeId], outcome: 1});
+                        if (startNode !== undefined)
+                          acquired.push(...propagateAcquix(allKnowledge, startNode, -1));
+                      } else {
+                        history.push({diff: difficultyOf[answer.challengeId], outcome: 0});
+                        if (startNode !== undefined)
+                          notAcquired.push(...propagateAcquix(allKnowledge, startNode, 1));
+                      }
+                    }
+                  });
+
+                  const pixScore = Math.round(64 * acquired.length / Object.keys(knowledgeOf).length);
+
+                  let estimatedLevel = 4;
+                  let minScore = 1000;
+                  if (history.length > 0) { 
+                    for(let level = 0; level <= 6; level += 0.5) {
```

Comment:
Je sais mais faut que je fasse une itération sur les valeurs de 0 à 7 avec un pas de 0,5 et si je fais ça avec _.times le code sera illisible pour quelqu'un qui ne connait pas lodash. Là le for tout le monde le comprend.